src/Main.hs

```diff
@@ -38,9 +45,94 @@ printIssue = BSL.putStr . (<> "\0") . encode
 
 --------------------------------------------------------------------------------
 
-shFiles :: [FilePath] -> IO [FilePath]
-shFiles paths =
-  fmap concat $! sequence $! fmap (matched . globDir [compile "**/*.sh"]) paths
+shellScripts :: [FilePath] -> IO [FilePath]
+shellScripts paths = do
+  dotShFiles <- concat . fst <$> globDir patterns "."
+  allScripts <- filterM validateScript $ dotShFiles ++ otherFiles
+  return $ fmap clean allScripts
   where
-    matched :: Functor f => f ([[a]], [b]) -> f [a]
-    matched x = (concat . fst) <$> x
+    (dirs, otherFiles) = partition hasTrailingPathSeparator paths
+
+    clean :: String -> String
+    clean ('.' : '/' : x) = x
+    clean x               = x
+
+    patterns :: [Pattern]
+    patterns = fmap (compile . (++ "**/*.sh")) dirs
+
+    validateScript :: FilePath -> IO Bool
+    validateScript x = doesFileExist x &&^ isShellScript x
+
+--------------------------------------------------------------------------------
+
+-- | Determines whether a file is a shell script that we can work with.
+isShellScript :: FilePath -> IO Bool
+isShellScript path =
+  if hasExtension path
+    then return hasShellExtension
+    else do
+      header <- readHeader
+      if hasShebang header
```

Comment:
@mrb I suspect that this could be optimised by simply using the `shebang` parser rather than checking the header for `#!`.